### PR TITLE
Add the check condition to SCA stored data

### PR DESCRIPTION
**Pull request by carlocas**
|Related issue|
|-------------|
[3598](https://github.com/wazuh/wazuh/issues/3598)

## Description
The purpose of this PR is that the field `condition` of the module sca,  should be stored in the database.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation
- [x] Source upgrade
- Memory tests
  - [x] Valgrind report for affected components
    - [x] wazuh-modulesd

```
==12099== HEAP SUMMARY:
==12099==     in use at exit: 3,277 bytes in 39 blocks
==12099==   total heap usage: 3,290 allocs, 3,251 frees, 1,134,512 bytes allocated
==12099== 
==12099== LEAK SUMMARY:
==12099==    definitely lost: 0 bytes in 0 blocks
==12099==    indirectly lost: 0 bytes in 0 blocks
==12099==      possibly lost: 0 bytes in 0 blocks
==12099==    still reachable: 3,277 bytes in 39 blocks
==12099==         suppressed: 0 bytes in 0 blocks
==12099== Reachable blocks (those to which a pointer was found) are not shown.
==12099== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==12099== 
==12099== For counts of detected and suppressed errors, rerun with: -v
==12099== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
  - [x] wazuh-analysisd
```
==16672== LEAK SUMMARY:
==16672==    definitely lost: 100 bytes in 9 blocks
==16672==    indirectly lost: 0 bytes in 0 blocks
==16672==      possibly lost: 0 bytes in 0 blocks
==16672==    still reachable: 12,296 bytes in 295 blocks
==16672==         suppressed: 0 bytes in 0 blocks
==16672== Reachable blocks (those to which a pointer was found) are not shown.
==16672== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==16672== 
==16672== For counts of detected and suppressed errors, rerun with: -v
==16672== ERROR SUMMARY: 5 errors from 5 contexts (suppressed: 0 from 0)
```
  - [x] wazuh-db
```
==11115== HEAP SUMMARY:
==11115==     in use at exit: 8,344 bytes in 2 blocks
==11115==   total heap usage: 32 allocs, 30 frees, 64,140 bytes allocated
==11115== 
==11115== LEAK SUMMARY:
==11115==    definitely lost: 0 bytes in 0 blocks
==11115==    indirectly lost: 0 bytes in 0 blocks
==11115==      possibly lost: 0 bytes in 0 blocks
==11115==    still reachable: 8,344 bytes in 2 blocks
==11115==         suppressed: 0 bytes in 0 blocks
==11115== Reachable blocks (those to which a pointer was found) are not shown.
==11115== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==11115== 
==11115== For counts of detected and suppressed errors, rerun with: -v
==11115== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
  - [x] No error found with scan-build
  - [x] No error found with address-sanitizer

- Custom tests
  - [x] Check if the field `condition` is show in the database
Now, in the database show the following message:
```
CREATE TABLE sca_check (   scan_id INTEGER REFERENCES sca_scan_info (id),   id INTEGER PRIMARY KEY,   policy_id TEXT REFERENCES sca_policy (id),   title TEXT,   description TEXT,   rationale TEXT,   remediation TEXT,   file TEXT,   process TEXT,   directory TEXT,   registry TEXT,   command TEXT,   `references` TEXT,   result TEXT,   `status` TEXT,   reason TEXT, condition TEXT);
CREATE INDEX policy_id_index ON sca_check (policy_id);
```
```
"timestamp": "2019-07-15T10:40:01.218+0200",
  "rule": {
    "level": 3,
    "description": "System audit for web-related vulnerabilities: Web exploits: '.yop' is an uncommon file name inside htdocs - Possible compromise",
    "id": "19009",
    "firedtimes": 5,
    "mail": false,
    "groups": [
      "sca"
    ],
    "gdpr": [
      "IV_35.7.d"
    ]
  },
  "agent": {
    "id": "000",
    "name": "host"
  },
  "manager": {
    "name": "host"
  },
  "id": "1563180001.8174",
  "decoder": {
    "name": "sca"
  },
  "data": {
    "sca": {
      "type": "check",
      "scan_id": "1237957229",
      "policy": "System audit for web-related vulnerabilities",
      "check": {
        "id": "1004",
        "title": "Web exploits: '.yop' is an uncommon file name inside htdocs - Possible compromise",
        "compliance": {
          "pci_dss": "6.5, 6.6, 11.4"
        },
        "condition": "any",
        "directory": [
          "/var/www",
          "/var/htdocs",
          "/home/httpd",
          "/usr/local/apache",
          "/usr/local/apache2",
          "/usr/local/www"
        ],
        "status": "Not applicable",
        "reason": "Could not open '/var/www': No such file or directory"
      }
    }
  },
  "location": "sca"
```
  - [x] Check if the field `condition` is shown when we upgrade the manager and the agent.
First, we obtain a message output from the database of version 3.9. 
```
928022162|5041|cis_debian|Ensure root is the only UID 0 account|Any account with UID 0 has superuser privileges on the system.|This access must be limited to only the default root account and only from the system console. Administrative access must be through an unprivileged account using an approved mechanism as noted in Item 5.6 Ensure access to the su command is restricted.|Remove any users other than root with UID 0 or assign them a new UID if appropriate.|/etc/passwd||||||passed|
```
Upgrade the manager and check the message. We can see that the field `condition` is now visible.
```
928022162|5041|cis_debian|Ensure root is the only UID 0 account|Any account with UID 0 has superuser privileges on the system.|This access must be limited to only the default root account and only from the system console. Administrative access must be through an unprivileged account using an approved mechanism as noted in Item 5.6 Ensure access to the su command is restricted.|Remove any users other than root with UID 0 or assign them a new UID if appropriate.|/etc/passwd||||||passed|||any
```
Connect an agent with the version 3.9 and, check if the field ´condition´ is shown. 
```
1510267933|6574|cis_rhel7|Disable standard boot services - NetFS Enabled||||||/etc/rc.d/rc2.d,/etc/rc.d/rc3.d,/etc/rc.d/rc4.d,/etc/rc.d/rc5.d||||passed|||
```
The space to the field is empty, now upgrade the agent, and check if the database has been updated. 
```
1585019112|6574|cis_rhel7|Disable standard boot services - NetFS Enabled||||||/etc/rc.d/rc2.d,/etc/rc.d/rc3.d,/etc/rc.d/rc4.d,/etc/rc.d/rc5.d||||passed|||any
```

### CODE REVIEW

File: extensions/elasticsearch/6.x/wazuh-template.json [hunk 1]
```diff
@@ -415,6 +415,9 @@
                 "check.rationale": {
                   "type": "text"
                 },
+                "check.condition": {
```
Comment (chemamartinez) [inline extensions/elasticsearch/6.x/wazuh-template.json]: This field is inside the `oscap` block, it has no relation to SCA but the integration with OpenSCAP.

File: extensions/elasticsearch/7.x/wazuh-template.json [hunk 2]
```diff
@@ -416,6 +416,9 @@
               "check.rationale": {
                 "type": "text"
               },
+              "check.condition": {
```
Comment (chemamartinez) [inline extensions/elasticsearch/7.x/wazuh-template.json]: This field is inside the `oscap` block, it has no relation to SCA but the integration with OpenSCAP.

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 3]
```diff
@@ -1359,6 +1360,13 @@ static int CheckEventJSON(cJSON *event,cJSON **scan_id,cJSON **id,cJSON **name,c
             return retval;
         }
 
+        *condition = cJSON_GetObjectItem(*check, "condition");
+        
+        obj= *condition;
+        if( obj && !obj->valuestring){
+            merror ("Malformed JSON: field 'condition' must be a string");
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: If the check fails, it has to return. Please add `return retval;`

File: src/wazuh_db/schema_upgrade_v3.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+BEGIN;
```
Comment (chemamartinez) [inline src/wazuh_db/schema_upgrade_v3.sql]: Every new file has to contain a header including the Wazuh copyright among other things.

File: src/wazuh_db/wdb_upgrade.c [hunk 5]
```diff
@@ -40,6 +40,13 @@ wdb_t * wdb_upgrade(wdb_t *wdb) {
         }
         /* Fallthrough */
     case 2:
+    mdebug2("Updating database for agent %s to version 3", wdb->agent_id);
```
Comment (chemamartinez) [inline src/wazuh_db/wdb_upgrade.c]: Format this message to the same level as the rest inside the `case` block.

File: framework/wazuh/tests/data/schema_sca_test.sql [hunk 6]
```diff
@@ -51,146 +52,146 @@ CREATE TABLE sca_check_rules (id_check INTEGER REFERENCES sca_check (id), type T
...
+INSERT INTO sca_check (scan_id,id,policy_id,title,description,rationale,remediation,file,process,directory,registry,command,"references","result",status,reason,condition) VALUES (
+339510093,1000,'system_audit','PHP - Register globals are enabled',NULL,NULL,NULL,'/etc/php.ini',NULL,NULL,NULL,NULL,NULL,'','Not applicable','File /etc/php.ini not found',any);
+INSERT INTO sca_check (scan_id,id,policy_id,title,description,rationale,remediation,file,process,directory,registry,command,"references","result",status,reason,condition) VALUES (
+339510093,1001,'system_audit','PHP - Expose PHP is enabled',NULL,NULL,NULL,'/etc/php.ini',NULL,NULL,NULL,NULL,NULL,'','Not applicable','File /etc/php.ini not found',any);
+INSERT INTO sca_check (scan_id,id,policy_id,title,description,rationale,remediation,file,process,directory,registry,command,"references","result",status,reason,condition) VALUES (
+339510093,1002,'system_audit','PHP - Allow URL fopen is enabled',NULL,NULL,NULL,'/etc/php.ini',NULL,NULL,NULL,NULL,NULL,'','Not applicable','File /etc/php.ini not found',)any;
```
Comment (chemamartinez) [inline framework/wazuh/tests/data/schema_sca_test.sql]: Check this typo `,)any;`

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 7]
```diff
@@ -36,16 +36,16 @@ static int SaveEventcheck(Eventinfo *lf, int exists, int *socket, int id , int s
...
-static int CheckEventJSON(cJSON *event,cJSON **scan_id,cJSON **id,cJSON **name,cJSON **title,cJSON **description,cJSON **rationale,cJSON **remediation,cJSON **compliance,cJSON **check,cJSON **reference,cJSON **file,cJSON **directory,cJSON **process,cJSON **registry,cJSON **result,cJSON **status,cJSON **reason,cJSON **policy_id,cJSON **command, cJSON **rules);
-static int CheckPoliciesJSON(cJSON *event,cJSON **policies);
-static int CheckDumpJSON(cJSON *event,cJSON **elements_sent,cJSON **policy_id,cJSON **scan_id);
-static void FillCheckEventInfo(Eventinfo *lf,cJSON *scan_id,cJSON *id,cJSON *name,cJSON *title,cJSON *description,cJSON *rationale,cJSON *remediation,cJSON *compliance,cJSON *reference,cJSON *file,cJSON *directory,cJSON *process,cJSON *registry,cJSON *result,cJSON *status,cJSON *reason,char *old_result,cJSON *command);
-static void FillScanInfo(Eventinfo *lf,cJSON *scan_id,cJSON *name,cJSON *description,cJSON *pass,cJSON *failed,cJSON *invalid,cJSON *total_checks,cJSON *score,cJSON *file,cJSON *policy_id);
+static int SavePolicyInfo(Eventinfo *lf, int *socket, char *name,char *file, char * id,char *description,char * references, char *hash_file);
```
Comment (JuantAldea) [inline src/analysisd/decoders/security_configuration_assessment.c]: The scope of initial request was only the lines you had modified. But now that you have changed many, please fix every prototype, and verify that every param has been corrected.

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 8]
```diff
@@ -735,7 +736,7 @@ static void HandleCheckEvent(Eventinfo *lf,int *socket,cJSON *event) {
 
     mdebug1("Checking event JSON fields");
 
-    if(!CheckEventJSON(event,&scan_id,&id,&name,&title,&description,&rationale,&remediation,&compliance,&check,&reference,&file,&directory,&process,&registry,&result,&status,&reason,&policy_id,&command,&rules)) {
+    if(!CheckEventJSON(event, &scan_id, &id, &name, &title, &description, &rationale, &remediation, &compliance, &condition, &check, &reference, &file, &directory, &process, &registry, &result, &status, &reason, &policy_id, &command, &rules)) {
```
Comment (JuantAldea) [inline src/analysisd/decoders/security_configuration_assessment.c]: Split this into multiple lines like
image
In general try to keep lines ~140 max, but don't hesitate to keep them shorter if you have to split them anyway.

The same applies to many changed lines of this PR.

File: src/wazuh_db/wdb_sca.c [hunk 9]
```diff
@@ -135,7 +135,8 @@ int wdb_sca_find(wdb_t * wdb, int pm_id, char * output) {
 }
 
 /* Insert configuration assessment entry. Returns 0 on success or -1 on error (new) */
-int wdb_sca_save(wdb_t * wdb, int id,int scan_id,char * title,char *description,char *rationale,char *remediation, char * file,char * directory,char * process,char * registry,char * reference,char * result,char * policy_id,char * command,char *status,char *reason) {
+int wdb_sca_save(wdb_t * wdb, int id, int scan_id, char * title, char *description, char *rationale, char *remediation, char *condition, char * file,
```
Comment (JuantAldea) [inline src/wazuh_db/wdb_sca.c]: Seize the opportunity to reformat all parameters to `type *variable`, removing the space between the `*` and the variable.
Do the same if find the same inconsistence in any line changed by this PR.

File: extensions/elasticsearch/6.x/wazuh-template.json [hunk 10]
```diff
@@ -1195,6 +1195,10 @@
                         }
                       }
                     },
+                    "condition": {
```
Comment (chemamartinez) [inline extensions/elasticsearch/6.x/wazuh-template.json]: Remove this field from the template, the `condition` field won't be shown as an alert field, same as `rule`.

**Pull request merged**